[CMake] Use standard macros + fix out-of-tree builds

guparan opened the pull request:
- Use `sofa_add_module` intensively for all new NG modules.
- Use `sofa_add_collection` for all old modules that are now simple collections.
- Make Sofa.GUI a real module.
- Fix disabled targets inclusion in SDK

:warning: This PR changes basically all CMake option names. E.g. `PLUGIN_SOFAGENERALLOADER` becomes `COLLECTION_SOFAGENERALLOADER`. The good thing is that it's way clearer.

This PR also contains the same fix as #3103

EDIT 
New changes:
- Move Sofa.GL module under Sofa directory
- SofaKernel is gone.
- Everything was moved under applications/collections/deprecated
- New cosmetic feature in SofaMacros to hide sub-options when base-option is disabled (like cmake_dependent_option but simpler). E.g. when I disable SOFA_COMPONENT, all SOFA_COMPONENT_* are hidden.
- A bit of cleaning

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

guparan committed:
Use standard macros

guparan committed:
Use standard macros (2)

guparan committed:
Use standard macros (3)

guparan committed:
FIX Collision.Response.Mapper test option

guparan committed:
[Sofa.GUI] Clean CMakeLists

guparan committed:
[CMake] FIX SceneCreator option

guparan committed:
[Sofa.Simulation] Clean CMakeLists

guparan committed:
[CMake] Legacy modules are now collections

guparan committed:
[CMake] Legacy modules are now collections (2)

guparan committed:
[Sofa.GUI] Make it a module

guparan committed:
[CMake] FIX all modules with optional targets

guparan committed:
Clean SofaMacros

guparan committed:
[Sofa.Simulation] Clean CMakeLists

guparan committed:
[SofaMacros] Remove sofa_add_component_subdirectory

alxbilger commented:
Does it mean that #3103 should be closed?

(Note that the build is failing)

fredroy commented on CMakeLists.txt:
```diff
@@ -159,86 +158,68 @@ endif()
 option(SOFA_WITH_OPENGL "Compile Sofa with OpenGL support. This will define the SOFA_WITH_OPENGL and SOFA_NO_OPENGL CMake variables." ON)
 if(SOFA_WITH_OPENGL)
-    add_subdirectory(${SOFA_KERNEL_SOURCE_DIR}/modules/Sofa.GL)
+    sofa_add_module(${SOFA_KERNEL_SOURCE_DIR}/modules/Sofa.GL Sofa.GL ON)
```
Sofa.GL is supposed to be enabled/built before Sofa.GUI could be activated, as some packages inside Sofa.GL needs OpenGL

fredroy replied:
Maybe this is the opportunity to move simply (Sofa.)GL into the Sofa/ folder (similarly to (Sofa,)GUI)

guparan replied:
Done.

fredroy reviewed (changes requested):
I am a bit sad that we lost simplicity in cmake files 😢. 
I liked the fact that Sofa/framework was (supposed to?) be buildable by iteelf; and that the root SOFA CMakefile was kind of slimmer/less complicated and actually was (supposed to be ?) not mandatory.

guparan commented:
> I am a bit sad that we lost simplicity in cmake files cry. I liked the fact that Sofa/framework was (supposed to?) be buildable by iteelf; 

How? By unchecking the SOFA_COMPONENT option? It's still possible this way. The thing is that Framework is (unfortunately) not a module and thus should not be treated as one if we want to be consistent. I'm not against reverting the framework/CMakeLists.txt deletion though.

> and that the root SOFA CMakefile was kind of slimmer/less complicated and actually was (supposed to be ?) not mandatory.

This file was and still is mandatory. At least, I think it's clearer now with the changes I propose because we see what is directly below the root CMakeLists (all the remaining add_subdirectory calls) and what could be done from somewhere else. Ideally we should just have a some options/variables and a bunch of add_subdirectory calls in this file.

guparan committed:
[SofaMacros] Add mechanism to hide suboptions when parent is disabled

guparan committed:
[Sofa.GL] Move the module under Sofa directory

guparan committed:
[extlibs] Move SofaKernel/extlibs to extlibs

guparan committed:
[collections] Move all old modules and meta-modules to collections/deprecated

guparan committed:
[CMake] Clean things

guparan committed:
[collections] Move SofaComponentAll to collections/deprecated + Clean stuff

guparan commented:
I pushed a lot of new changes and file moves.
I'm ok to split the PR in 2 if necessary.

guparan commented:
New changes:
- Move Sofa.GL module under Sofa directory
- SofaKernel is gone.
- Everything was moved under applications/collections/deprecated
- New cosmetic feature in SofaMacros to hide sub-options when base-option is disabled (like cmake_dependent_option but simpler). E.g. when I disable SOFA_COMPONENT, all SOFA_COMPONENT_* are hidden.
- A bit of cleaning

guparan committed:
[CMake] Use dependent options for cleaner CMake-GUI

guparan commented:
> Does it mean that #3103 should be closed?
> 
> (Note that the build is failing)

Yep, doing it right now.

guparan commented:
https://github.com/guparan/sofa/blob/cmake_ng_use_standard_macros/Sofa/GUI/src/sofa/gui/init.cpp should not be a plugin

guparan committed:
[CMake] CLEAN deprecation messages

Deprecation message for old modules will not be printed if the find_package has been done in SOFA.

fredroy committed:
compat modules were not acitvated even if legacy_headers was on

fredroy commented:
[ci-build][force-full-build][with-all-tests]

guparan committed:
[Sofa.Simulation] FIX init.cpp for non-plugin library

guparan committed:
[Sofa.GUI] FIX init.cpp for non-plugin library

guparan committed:
[SofaMacros] New macro sofa_add_subdirectory to refactor all sofa_add_*

guparan committed:
[CMake] Use new sofa_add_subdirectory macro

guparan committed:
[SofaMeshCollision] Clean CMakeLists

guparan committed:
[SofaMacros] Only add modules and plugins to plugins_list.conf

guparan commented:
[ci-build][force-full-build][with-all-tests]

guparan committed:
[CMake] FIX Sofa.Compat and Sofa.Component.Compat are libraries, not modules

guparan committed:
[CMake] FIX Sofa.Config is a library, not a module

guparan commented:
[ci-build][force-full-build][with-all-tests]

guparan commented:
Here we are. Ready to re-review @fredroy :wink:

fredroy committed:
Merge branch 'master' into cmake_ng_use_standard_macros

fredroy reviewed (approved):
I guess I would need more info with a sofadev meeting but unfortunately I wont be able to attend this week 😢
as far as I understood you introduce a new macro defining the type of the project, which looks good to me (and defined library/collection/module/plugin as we said) and removed the _external suffix. 
Looks good to me as far as I can see ! (even if I would have wished you keep the Sofa/framework/CMakeLists.txt :p )

guparan commented:
> I guess I would need more info with a sofadev meeting but unfortunately I wont be able to attend this week cry as far as I understood you introduce a new macro defining the type of the project, which looks good to me (and defined library/collection/module/plugin as we said) and removed the _external suffix. Looks good to me as far as I can see ! (even if I would have wished you keep the Sofa/framework/CMakeLists.txt :p )

Exactly, I refactored the sofa_add_* macros under one sofa_add_subdirectory macro. The reason is that I didn't want to create a misleading sofa_add_library macro. I like the fact that sofa_* macro have the same name than the standard CMake macro they override.
I'll resolve the conflicts and revert the framework/CMakeLists removal :wink:

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into cmake_ng_use_standard_macros

# Conflicts:
#	Sofa/GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.cpp
#	Sofa/GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglLineAxis.cpp
#	Sofa/framework/Config/cmake/SofaMacrosConfigure.cmake

guparan committed:
[framework] Revert framework/CMakeLists.txt removal

guparan commented:
Here you go @fredroy

guparan commented:
> Several points here:
> 
>     * this PR might be breaking for plugin compilation (including move of files)

No it is not. The only change that could be called "breaking" is the renaming of the CMake options to enable/disable libraries and modules. However, those options keep the same default behavior so it will only affect people who build their own SOFA with custom options. New option names are very close to old option name so it's pretty easy to do the transition. 

> 
>     * this PR changes the way we organize / name the code (which is good!)
> 
> 
> It would be therefore nice to rediscuss it with you both @guparan and @fredroy to all agree on the organization and merge it, but a question comes: does this seem realistic to merge it before the release?

Yes because having everything called PLUGIN_* is a nightmare.
